### PR TITLE
Fix #4976. Node "Voronoi on Solid" produce incorrect results.

### LINKED ISSUE #4976: Node "Voronoi on Solid" produce incorrect results.
(for Pull Request)
Sverchok 1.3.0-alpha/master, Windows 11, Blender 3.6.1.

image

Strange corners on Surface mode:

image

### PR DESCRIPTION
**Pull request by satabol**
fix #4976. Update node "Voronoi on Solid":
- update using Voronoi. Get new cells for border cells.
- update parameter inset (now it use distance. Before was percent.)

Description of fix:

Source algorithm of node "Voronoi on Solid" try to use closed cells of voronoi (1) for source sites but not all sites are closed. So old algorithm add bounding box points to sites (2):
image
This produce closed sites that was opened but this is produce some disturbing actions to source sites and some areas of bounding box are still opened. This can produce some glitches:
image

In other cases it can produce other glitches (view of Voronoi sites):
image
Result look unpredictable:
image

So I have a proposal:
1. Do not use inset as percentage
2. Modify algorithm of getting voronoi closed cells to get cells that fill bounding box of source solid

Combine 1 and 2 with function voronoi_on_mesh_bmesh. Bounding box exceed solid bounding box for 1 in all directions:
image

and now results looks good.

### CODE REVIEW

File: nodes/spatial/voronoi_on_solid_mk2.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        default = 0.1,
+        description="Distance to leave between generated Voronoi regions",
+        update = updateNode)
+
+    scale_types = [
+            ('SITE', "Site", "Scale each region relative to corresponding site location", 0),
+            ('MEAN', "Barycenter", "Scale each region relative to it's barycenter, i.e. average location of it's vertices", 1)
+        ]
+
+    scale_center : EnumProperty(
```
Comment (portnov) [inline nodes/spatial/voronoi_on_solid_mk2.py:75]: this is not used anymore as far as I understand?

File: nodes/spatial/voronoi_on_solid_mk2.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        self.outputs.new('SvSolidSocket', "InnerSolid")
+        self.outputs.new('SvSolidSocket', "OuterSolid")
+
+    def draw_buttons(self, context, layout):
+        layout.prop(self, "mode")
+        layout.prop(self, "flat_output")
+
+    def draw_buttons_ext(self, context, layout):
+        self.draw_buttons(context, layout)
+        layout.prop(self, 'scale_center')
```
Comment (portnov) [inline nodes/spatial/voronoi_on_solid_mk2.py:101]: and this also.

File: nodes/spatial/voronoi_on_solid_mk2.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        layout.prop(self, 'accuracy')
+
+    def scale_cells(self, verts, sites, insets, precision):
```
Comment (portnov) [inline nodes/spatial/voronoi_on_solid_mk2.py:104]: and this?

File: nodes/spatial/voronoi_on_solid_mk2.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            for solid, sites, inset in zip_long_repeat(*params):
+                #verts, edges, faces = voronoi_on_solid(solid, sites, do_clip=True, clipping=None)
+                # see more info: https://github.com/nortikin/sverchok/pull/4977
+                box = solid.BoundBox
+                clipping = 1
+                x_min, x_max = box.XMin - clipping, box.XMax + clipping
+                y_min, y_max = box.YMin - clipping, box.YMax + clipping
+                z_min, z_max = box.ZMin - clipping, box.ZMax + clipping
+                bounds = list(itertools.product([x_min,x_max], [y_min, y_max], [z_min, z_max]))
+                verts, edges, faces = voronoi_on_mesh_bmesh(bounds, [ [0,1,3,2], [2,3,7,6], [6,7,5,4], [4,5,1,0], [2,6,4,0], [7,3,1,5] ], len(sites), sites, spacing=inset, mode='VOLUME' )
```
Comment (portnov) [inline nodes/spatial/voronoi_on_solid_mk2.py:155]: numbers in brackets look magic :)
I'd suggest to assign them to a separate variable and add a comment about what is this.

File: nodes/spatial/voronoi_on_solid_mk2.py (hunk 1 above)
Comment (satabol) [reply nodes/spatial/voronoi_on_solid_mk2.py:75]: > this is not used anymore as far as I understand?

Yes. It is not used more.

Comment (satabol) [reply nodes/spatial/voronoi_on_solid_mk2.py]: fixed

File: nodes/spatial/voronoi_on_solid_mk2.py (hunk 2 above)
Comment (satabol) [reply nodes/spatial/voronoi_on_solid_mk2.py]: fixed. removed.

File: nodes/spatial/voronoi_on_solid_mk2.py (hunk 3 above)
Comment (satabol) [reply nodes/spatial/voronoi_on_solid_mk2.py]: fixed. scale_center removed.

File: nodes/spatial/voronoi_on_solid_mk2.py (hunk 4 above)
Comment (satabol) [reply nodes/spatial/voronoi_on_solid_mk2.py:123]: What brackets? If an argument for voronoi_on_mesh_bmesh then it is a cube faces that is a bounding box.

**Comment by satabol** [comment]
Description of fix:

image

**Comment by portnov** [comment]
now it's ok with me :)

**Pull request merged**